### ISSUE TITLE
Add functionality for multiple popups

### ISSUE THREAD

**Issue by zacplansky** [zacplansky/justunodollars#23]
Below is an outline of one simple, resilient approach to support multiple popups—each with its own design, triggers, and targeting. The goal is to change as little of the existing structure as possible, while introducing the concept of an array of popups in your plugin settings. From there, the frontend logic simply iterates over each defined popup and decides whether to show it.

Please read through this, and carefully implement. Think ultrahard about how to do it right.

⸻

1. Store multiple popup configs in an array (instead of one “global” popup)

Right now, everything revolves around a single set of data in justunodollars_settings:

'popup_desktop_html' => '...',
'popup_mobile_html' => '...',
'show_on_desktop'    => 'yes',
'show_on_mobile'     => 'yes',
'delay_seconds'      => 5,
'scroll_trigger'     => 50,
...

A simple approach is to store an array of popups. For example:

'popups' => array(
    array(
        'id'                => 'popup_1',
        'title'             => 'Main Newsletter Popup',
        'desktop_html'      => '...',
        'mobile_html'       => '...',
        'show_on_desktop'   => 'yes',
        'show_on_mobile'    => 'yes',
        'delay_seconds'     => 5,
        'scroll_trigger'    => 50,
        'repeat_delay_days' => 7,
        // etc...
    ),
    array(
        'id'                => 'popup_2',
        'title'             => 'Exit Intent Promo Popup',
        'desktop_html'      => '...',
        'mobile_html'       => '...',
        'exit_intent'       => 'yes',
        'delay_seconds'     => 0,
        'scroll_trigger'    => 0,
        'repeat_delay_days' => 3,
        // etc...
    ),
    // More popups...
),

Your existing single “popup” settings become fields inside each popup’s sub-array.

Key point: If you do this, you’ll want separate “seen” cookies / localStorage keys for each popup so that closing one popup doesn’t prevent others from appearing.

⸻

Minimal changes for the Install class

In JustUnoDollars_Install::activate(), instead of setting default HTML fields at the top level, you might do something like:

$default_settings = array(
    'popups' => array(
        array(
            'id'                => 'popup_1',
            'title'             => 'Default Popup',
            'desktop_html'      => self::get_default_popup_html('desktop'),
            'mobile_html'       => self::get_default_popup_html('mobile'),
            'delay_seconds'     => 5,
            'scroll_trigger'    => 50,
            'repeat_delay_days' => 7,
            'show_on_mobile'    => 'yes',
            'show_on_desktop'   => 'yes',
            'exit_intent'       => 'yes',
            // plus other fields
        )
    ),
    'funnelkit_webhook_url' => '',
    // anything else that is at the "global" level
);

add_option('justunodollars_settings', $default_settings);

⸻

2. Admin UI changes: allow editing multiple popups

You have a single Popup Design tab that edits:

$settings['popup_desktop_html']
$settings['popup_mobile_html']

To support multiples, the simplest approach is to:
	1.	Change your “Popup Design” tab to list all popups currently stored in justunodollars_settings['popups'].
	2.	Provide an “Add New Popup” button that pushes a new array entry.
	3.	Let the admin pick which popup to edit (using a dropdown or sub-tab).
	4.	Show the same design fields and trigger fields, but store them in popups[index]['desktop_html'], etc.

Example in your render_popup_design_tab():

public function render_popup_design_tab() {
    $settings = $this->settings;
    $popups = isset($settings['popups']) && is_array($settings['popups']) ? $settings['popups'] : array();

    // Maybe get the "active" popup index from the query string
    $current_popup_index = isset($_GET['popup_index']) ? (int) $_GET['popup_index'] : 0;

    // Handle user selecting a different popup via <select> or a sub-tab
    if (!isset($popups[$current_popup_index])) {
        // Fall back if index is out of range
        $current_popup_index = 0;
    }

    $current_popup = $popups[$current_popup_index];

    // Show a simple dropdown to switch between popups
    echo '<h2>Popup Designs</h2>';
    echo '<select onchange="location = this.value;">';
    foreach ($popups as $index => $popup_config) {
        $selected = ($index === $current_popup_index) ? 'selected' : '';
        echo '<option '.$selected.' value="'.esc_url(add_query_arg(['popup_index' => $index])).'">';
        echo esc_html($popup_config['title'] ?? "Popup #$index");
        echo '</option>';
    }
    echo '</select>';

    // Now show fields for the "current" popup
    ?>
    <form method="post" action="options.php">
        <?php settings_fields('justunodollars_settings_group'); ?>

        <h3><?php echo esc_html($current_popup['title'] ?? 'Unnamed Popup'); ?></h3>

        <label for="justunodollars_popup_title">Popup Title</label><br>
        <input type="text" name="justunodollars_settings[popups][<?php echo $current_popup_index; ?>][title]"
            value="<?php echo esc_attr($current_popup['title'] ?? ''); ?>" /><br><br>

        <h4>Desktop HTML</h4>
        <textarea name="justunodollars_settings[popups][<?php echo $current_popup_index; ?>][desktop_html]" rows="15" cols="80" class="large-text code"><?php 
            echo esc_textarea($current_popup['desktop_html'] ?? ''); 
        ?></textarea>

        <h4>Mobile HTML</h4>
        <textarea name="justunodollars_settings[popups][<?php echo $current_popup_index; ?>][mobile_html]" rows="15" cols="80" class="large-text code"><?php 
            echo esc_textarea($current_popup['mobile_html'] ?? ''); 
        ?></textarea>

        

        <?php submit_button(__('Save Popup')); ?>
    </form>
    <?php
}

In your sanitize_settings() method, you’ll want to iterate over $input['popups'] and ensure each sub-array has the expected fields. If one is missing desktop_html, you fall back to the existing one, etc.

⸻

3. Frontend: iterate over each popup, decide whether to show it

The main changes come in JustUnoDollars_Frontend. Right now, you have:

// in constructor
$this->settings = get_option('justunodollars_settings', array());
...
add_action('wp_footer', array($this, 'add_popup_html'));

Inside add_popup_html(), you currently do:

if (!$this->should_show_popup()) {
    return;
}

echo '<div id="justunodollars-popup" ...>';
// desktop HTML
// mobile HTML
echo '</div>';

To handle multiple popups, you can store them each in separate DOM containers (with unique IDs), or wrap them in one container. For instance:
	1.	Loop over each popup in $this->settings['popups'].
	2.	For each popup, check if should_show_popup($popup_data) returns true.
	3.	If it should show, output that popup’s HTML in a <div id="justunodollars-popup-<?=$popup['id']?>">...</div> block.

Example (high-level pseudocode):

public function add_popup_html() {
    // If no popups or not an array, bail
    if (empty($this->settings['popups']) || !is_array($this->settings['popups'])) {
        return;
    }

    foreach ($this->settings['popups'] as $popup_config) {
        // Decide if we should show this particular popup
        if (!$this->should_show_this_popup($popup_config)) {
            continue;
        }
        
        // Output container
        $popup_id = esc_attr($popup_config['id']);
        echo '<div id="justunodollars-popup-' . $popup_id . '" class="justunodollars-popup" style="display:none;">';
        echo '<div class="justunodollars-overlay"></div>';

        // Desktop markup
        echo '<div class="justunodollars-modal desktop">';
        if (!empty($popup_config['desktop_html'])) {
            echo $this->kses_form_html($popup_config['desktop_html']);
        }
        echo '</div>';

        // Mobile markup
        echo '<div class="justunodollars-modal mobile">';
        if (!empty($popup_config['mobile_html'])) {
            echo $this->kses_form_html($popup_config['mobile_html']);
        }
        echo '</div>';

        echo '</div>'; // close .justunodollars-popup
    }
}

Then in your enqueue_scripts() or your JS logic, you also loop over each popup from the localized script data. You might send an array of objects—one object per popup:

wp_localize_script(
    'justunodollars-frontend',
    'JustUnoDollarsData',
    array(
        'ajax_url' => admin_url('admin-ajax.php'),
        'nonce'    => wp_create_nonce('justunodollars_nonce'),
        'popups'   => array_map(function($popup){
            return array(
                'id'                => $popup['id'],
                'delay_seconds'     => intval($popup['delay_seconds']),
                'scroll_trigger'    => intval($popup['scroll_trigger']),
                'repeat_delay_days' => intval($popup['repeat_delay_days']),
                'exit_intent'       => !empty($popup['exit_intent']) && $popup['exit_intent'] === 'yes',
                'show_on_mobile'    => $popup['show_on_mobile'] ?? 'yes',
                'show_on_desktop'   => $popup['show_on_desktop'] ?? 'yes',
                // etc.
            );
        }, $this->settings['popups'])
    )
);

Then in justunodollars-frontend.js, rather than a single init(), you would have an initPopup(popupConfig) function that deals with time delay, scroll triggers, etc. You’d iterate over JustUnoDollarsData.popups and run that logic for each.

The simplest approach is basically copy/paste the existing single-popup logic but inside a loop, where each loop references #justunodollars-popup-${popup.id}, sets up its own “seen” localStorage key, etc.:

JustUnoDollarsData.popups.forEach((popupConfig) => {
   initSinglePopup(popupConfig);
});

function initSinglePopup(cfg) {
   // Do the same stuff your single popup code does:
   // 1) grab DOM references by ID (#justunodollars-popup-cfg.id)
   // 2) check localStorage "justunodollars_seen_" + cfg.id
   // 3) attach event listeners, etc.
}

Each popup’s state (has it been shown yet, closed, etc.) is isolated by referencing popupConfig.id. This prevents closing “Popup #2” from hiding “Popup #1”.

⸻

4. Targeting logic: pass each popup config into should_show_popup()

You already have a method should_show_popup() that checks device, page targeting, etc. Adjust it to receive a $popup_config array:

private function should_show_this_popup($popup_config) {
    // 1. Device check
    if (wp_is_mobile()) {
        if (!isset($popup_config['show_on_mobile']) || $popup_config['show_on_mobile'] !== 'yes') {
            return false;
        }
    } else {
        if (!isset($popup_config['show_on_desktop']) || $popup_config['show_on_desktop'] !== 'yes') {
            return false;
        }
    }

    // 2. Page targeting
    // 3. UTM source
    // 4. Referring domain
    // ... same as before, but referencing $popup_config for each condition

    return true;
}

Then your add_popup_html() calls should_show_this_popup($popup_config) for each popup.

⸻

5. Summary of the minimal refactor steps
	1.	Install defaults: In activate(), store an array popups => [...] instead of single top-level popup fields.
	2.	Admin editing:
	•	Add a loop or multiple sub-tabs to let user pick which popup to edit.
	•	Store each popup’s data in popups[index][field].
	•	In sanitize_settings(), ensure you handle sub-arrays properly.
	3.	Frontend:
	•	In add_popup_html(), loop over $this->settings['popups'] and output separate containers for each popup design.
	•	Update your localize script call to pass an array of popup configs to the JS.
	•	In your justunodollars-frontend.js, do forEach popup, run the triggers/time-delay logic. Use the popup.id to isolate each popup’s localStorage, DOM, etc.
	•	If you want each popup’s analytics tracked separately, pass the popup ID to track_event() so you know which popup is being “opened” or “closed.”

This approach keeps the plugin’s existing architecture mostly intact—no extra tables or big data structures needed. You’re simply turning your single “settings” object into an array of smaller “popup settings” objects, then looping in both the admin UI and front-end.

⸻

Additional Tips
	•	Naming: You might rename your existing fields from popup_desktop_html to desktop_html, etc., inside the new array.
	•	Analytics: If you want to track each popup individually in your _tracking_events table, add a column like popup_id VARCHAR(50), and pass that to log_tracking_event().
	•	Performance: For 2–3 popups, a forEach loop in JS is perfectly fine. You’re not likely to have dozens or hundreds of popups.

Once you do this refactor, you’ll have a robust foundation to create an unlimited number of popups, each with unique targeting rules.
